File: backend/src/main/java/com/mini/credit/service/DemandeCreditService.java
package com.mini.credit.service;

import com.mini.credit.dto.credit.AnalyseRisqueRequest;
import com.mini.credit.dto.credit.CreditValidationResult;
import com.mini.credit.dto.credit.DemandeCreditCreateRequest;
import com.mini.credit.dto.credit.DemandeCreditResponse;
import com.mini.credit.dto.credit.FraisCreditAEncaisserResponse;
import com.mini.credit.dto.credit.PreAnalyseRequest;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;

public interface DemandeCreditService {
    DemandeCreditResponse create(DemandeCreditCreateRequest request);
    DemandeCreditResponse getById(Long id);
    Page<DemandeCreditResponse> getAll(Pageable pageable);
    List<DemandeCreditResponse> getAll();
    Page<DemandeCreditResponse> getByMembre(Long membreId, Pageable pageable);  // PHASE 3B: Added Pageable
    List<DemandeCreditResponse> getByMembre(Long membreId);
    DemandeCreditResponse ajouterAnalyse(Long demandeId, AnalyseRisqueRequest request);
    DemandeCreditResponse preAnalyser(Long demandeId, String commentaire);
    DemandeCreditResponse preAnalyserDecision(Long demandeId, PreAnalyseRequest request);
    DemandeCreditResponse enregistrerObservationRisque(Long demandeId, String commentaire);
    DemandeCreditResponse validerAnalyseRisque(Long demandeId, String commentaire);
    DemandeCreditResponse controlerRisque(Long demandeId, String commentaire);
    DemandeCreditResponse controlerGarantie(Long demandeId, String commentaire);
    DemandeCreditResponse rejeter(Long demandeId, String commentaire);
    boolean isCurrentUserRequest(Long demandeId);
    boolean isCurrentUserMembre(Long membreId);
    List<DemandeCreditResponse> getCurrentMemberRequests();
    List<FraisCreditAEncaisserResponse> getFraisCreditAEncaisser();

    /**
     * PHASE 4: Valide les critères strictes de crédit (frais, garantie, analyse terrain)
     */
    CreditValidationResult validerCredit(Long demandeId);
}

File: frontend/src/app/features/credit/services/demande-credit.service.ts
import { inject, Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';

import { API_BASE_URL } from '../../../core/services/api.config';
import { DemandeCreditCreateRequest } from '../models/demande-credit-create-request';
import { DemandeCreditResponse } from '../models/demande-credit-response';
import { FraisCreditAEncaisserResponse } from '../models/frais-credit-a-encaisser-response';
import { AnalyseRisqueRequest } from '../models/analyse-risque-request';
import { PreAnalyseRequest } from '../models/pre-analyse-request';
import { Page } from '../../../shared/models/page.model';

@Injectable({
  providedIn: 'root'
})
export class DemandeCreditService {
  private readonly http = inject(HttpClient);
  private readonly baseUrl = `${API_BASE_URL}/demandes-credit`;

  create(request: DemandeCreditCreateRequest): Observable<DemandeCreditResponse> {
    return this.http.post<DemandeCreditResponse>(this.baseUrl, request);
  }

  ajouterAnalyse(demandeId: number, request: AnalyseRisqueRequest): Observable<DemandeCreditResponse> {
    return this.http.post<DemandeCreditResponse>(`${this.baseUrl}/${demandeId}/analyse`, request);
  }

  preAnalyser(demandeId: number, commentaire?: string): Observable<DemandeCreditResponse> {
    const params = commentaire ? new HttpParams().set('commentaire', commentaire) : undefined;
    return this.http.post<DemandeCreditResponse>(`${this.baseUrl}/${demandeId}/pre-analyse`, {}, { params });
  }

  preAnalyserDecision(demandeId: number, request: PreAnalyseRequest): Observable<DemandeCreditResponse> {
    return this.http.post<DemandeCreditResponse>(`${this.baseUrl}/${demandeId}/pre-analyse/decision`, request);
  }

  controlerRisque(demandeId: number, commentaire?: string): Observable<DemandeCreditResponse> {
    return this.validerAnalyseRisque(demandeId, commentaire);
  }

  enregistrerObservationRisque(demandeId: number, commentaire: string): Observable<DemandeCreditResponse> {
    const params = new HttpParams().set('commentaire', commentaire);
    return this.http.post<DemandeCreditResponse>(`${this.baseUrl}/${demandeId}/analyse-risque/observation`, {}, { params });
  }

  validerAnalyseRisque(demandeId: number, commentaire?: string): Observable<DemandeCreditResponse> {
    const params = commentaire ? new HttpParams().set('commentaire', commentaire) : undefined;
    return this.http.post<DemandeCreditResponse>(`${this.baseUrl}/${demandeId}/analyse-risque/valider`, {}, { params });
  }

  controlerGarantie(demandeId: number, commentaire?: string): Observable<DemandeCreditResponse> {
    const params = commentaire ? new HttpParams().set('commentaire', commentaire) : undefined;
    return this.http.post<DemandeCreditResponse>(`${this.baseUrl}/${demandeId}/controle-garantie`, {}, { params });
  }

  rejeter(demandeId: number, commentaire: string): Observable<DemandeCreditResponse> {
    const params = new HttpParams().set('commentaire', commentaire);
    return this.http.post<DemandeCreditResponse>(`${this.baseUrl}/${demandeId}/rejeter`, {}, { params });
  }

  getById(id: number): Observable<DemandeCreditResponse> {
    return this.http.get<DemandeCreditResponse>(`${this.baseUrl}/${id}`);
  }

  /**
   * Get all credit requests with pagination support
   * @param page Page number (default 0)
   * @param size Page size (default 10)
   */
  getAll(page: number = 0, size: number = 10): Observable<Page<DemandeCreditResponse>> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());
    return this.http.get<Page<DemandeCreditResponse>>(this.baseUrl, { params });
  }

  getFraisCreditAEncaisser(): Observable<FraisCreditAEncaisserResponse[]> {
    return this.http.get<FraisCreditAEncaisserResponse[]>(`${this.baseUrl}/frais-a-encaisser`);
  }

  // PHASE 3B: Paginated version
  getByMembrePaginated(membreId: number, page: number = 0, size: number = 10): Observable<Page<DemandeCreditResponse>> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());
    return this.http.get<Page<DemandeCreditResponse>>(`${this.baseUrl}/membre/${membreId}`, { params });
  }

  // Backward compatibility: keep original method
  getByMembre(membreId: number): Observable<DemandeCreditResponse[]> {
    return this.http.get<DemandeCreditResponse[]>(`${this.baseUrl}/membre/${membreId}`);
  }

  /**
   * Get current member's credit requests
   */
  getMesDemandes(): Observable<DemandeCreditResponse[]> {
    return this.http.get<DemandeCreditResponse[]>(`${this.baseUrl}/mes-demandes`);
  }
}